[libc][arm] implement a basic setjmp/longjmp

nickdesaulniers opened the pull request:
Note: our baremetal arm configuration compiles this as
`--target=arm-none-eabi`, so this code is built in -marm mode. It could be
smaller with `--target=armv7-none-eabi -mthumb`. The assembler is valid ARMv5,
or THUMB2, but not THUMB(1). Unclear yet if we need to support THUMB(1); may
depend on whether downstream users are using llvm-libc's cmake.

nickdesaulniers committed:
[libc][arm] implement setjmp/longjmp

Note: our baremetal arm configuration compiles this as
`--target=arm-none-eabi`, so this code is built in -marm mode. It could be
smaller with `--target=armv7-none-eabi -mthumb`. The assembler is valid ARMv5,
or THUMB2, but not THUMB(1). Unclear yet if we need to support THUMB(1); may
depend on whether downstream users are using llvm-libc's cmake.

nickdesaulniers committed:
format

frobtech commented on libc/src/setjmp/arm/longjmp.cpp:
```diff
@@ -0,0 +1,31 @@
+
+[[gnu::naked]]
+LLVM_LIBC_FUNCTION(void, longjmp, (__jmp_buf * buf, int val)) {
+  asm("ldm.w r0!, {r4, r5, r6, r7, r8, r9, r10, r11, r12, lr}\n\t"
```
I think the norm is to use `__asm__`.  Consider using a raw string with natural newlines.

The `!` is not really useful here, since `r0` is not used again.

The syntax allows `{r4-r12, lr}` which IMHO is more readable.

It's unusual to use the `.w` suffix rather than letting the assembler decide what encoding to use.  I'd just use `ldm r0, {r4-r12, lr}` here.

nickdesaulniers replied:
> I think the norm is to use __asm__.

Ugly. No.

> Consider using a raw string with natural newlines.

Ah, good idea. 96d8f14ca1e83bc6af942a70b5a30fd410a67266

> The syntax allows {r4-r12, lr} which IMHO is more readable.
> It's unusual to use the .w suffix rather than letting the assembler decide what encoding to use. I'd just use ldm r0, {r4-r12, lr} here.

96d8f14ca1e83bc6af942a70b5a30fd410a67266

nickdesaulniers commented on libc/src/setjmp/arm/longjmp.cpp:
```diff
@@ -0,0 +1,30 @@
+
+#if !defined(LIBC_TARGET_ARCH_IS_ARM)
+#error "Invalid file include"
+#endif
```
These warnings (pattern copied from other setjmp impl's in tree) are dumb. You'll get an obvious compiler diagnostic without it pretty quick! I'm going to remove them.

nickdesaulniers replied:
77384336fb4e

frobtech commented on libc/src/setjmp/arm/setjmp.cpp:
```diff
@@ -0,0 +1,27 @@
+[[gnu::naked]]
+LLVM_LIBC_FUNCTION(int, setjmp, (__jmp_buf * buf)) {
+  asm("mov r12, sp\n\t"
+      "stm.w r0!, {r4, r5, r6, r7, r8, r9, r10, r11, r12, lr}\n\t"
```
Same issues here: `.w`,`!`, register range syntax, raw string, `#` prefix.

nickdesaulniers replied:
96d8f14ca1e83bc6af942a70b5a30fd410a67266

nickdesaulniers committed:
remove useless preprocessor warnings

frobtech commented on libc/include/llvm-libc-types/jmp_buf.h:
```diff
@@ -32,6 +32,9 @@ typedef struct {
 #endif
+#elif defined(__arm__)
+  // r4, r5, r6, r7, r8, r9, r10, r11, r12, lr
+  long opaque [10];
```
Extra space before `[` here.

This is already a problem for the x86-64 case, but it's not kosher to use plain identifiers here, or in any standard C header.  They need to be `__`-prefixed in case a user has `#define opaque syntax-error` before `#include <setjmp.h>`.  We should look into a clang-tidy check to enforce that.

nickdesaulniers replied:
fixed in https://github.com/llvm/llvm-project/pull/93220/commits/77384336fb4ef69a615b59fc0152ffdc74a4fb4e

frobtech reviewed (commented):
This can be addressed separately later, but I think we need to rethink the approach to setjmp/longjmp in libc.  In particular, I think implementations like this (and the `jmp_buf` ABI) cannot be universal per machine.  This is a great implementation for bare metal, but not necessarily correct or sufficient for any given operating system.  For POSIX systems, there's the issue of stack vs signal mask atomicity in (sig)longjmp.  For various OS ABIs, there are other ABI elements to be concerned with, like shadow-call-stack or safestack ABIs.  For various contexts there are additional concerns like pointer mangling.  For various more complete ABIs, there are non-integer registers that need to be considered.

nickdesaulniers committed:
raw string literals

nickdesaulniers commented on libc/src/setjmp/arm/longjmp.cpp:
```diff
@@ -0,0 +1,31 @@
+      "mov sp, r12\n\t"
+      "movs r0, r1\n\t"
+      "it eq\n\t"
+      "moveq r0, 1\n\t"
```
96d8f14ca1e83bc6af942a70b5a30fd410a67266

nickdesaulniers commented:
> not necessarily correct or sufficient for any given operating system.

Yes, these definitely can get way more complex.  Perhaps I should retitle the PR to "implement _simple_ setjmp/longjmp".

efriedma-quic commented:
> Unclear yet if we need to support THUMB(1); may depend on whether downstream users are using llvm-libc's cmake.

There are two issues here:

- Targets that support ARM-mode, but the user specified -mthumb for reduced codesize.  You can override the choice using `__attribute__((target("arm")))`.
- Targets that support neither ARM-mode nor Thumb2 (Cortex-M0 etc.).  This is a hardware limitation.

nickdesaulniers commented:
Oh, I forgot there's machines that ONLY support Thumb(1) such as cortex-m0. cc @smithp35 

I guess a fundamental question we should answer for llvm-libc is "we claim to support 32b ARM, but we don't specify which minimum target architecture."

I think it would be handy to draw the line at armv6t2+, or even just armv7, but others (@petrhosek @smithp35 ) may feel differently?

lntue commented on libc/src/setjmp/arm/longjmp.cpp:
```diff
@@ -0,0 +1,26 @@
+[[gnu::naked]]
+LLVM_LIBC_FUNCTION(void, longjmp, (__jmp_buf * buf, int val)) {
+  asm(R"(
+      ldm r0, {r4-r12, lr}
```
Is it possible to have (link to) documentations / comments to help newbies like me on these asm?

nickdesaulniers replied:
The definitive document on ARM assembler (32b and 64b) is called the Arm ARM.

https://developer.arm.com/documentation/ddi0487/latest/ is the stable link to the latest edition of the non-stable document.

AAPCS is also handy: https://github.com/ARM-software/abi-aa?tab=readme-ov-file#document-locations

https://nickdesaulniers.github.io/blog/2023/03/10/disambiguating-arm/

nickdesaulniers replied:
This code is restoring r4-r11, sp, and lr from the memory referred to by `buf`.

It then returns `1` if `val` was zero, else `val`.  Since `lr` was restored, it doesn't return to the caller of `longjmp`, but instead the caller of `setjmp`.

nickdesaulniers replied:
If you'd like a quick primer on ARM assembler, such as the calling convention, let's hop on a video call?

nickdesaulniers replied:
added comments PTAL

lntue commented on libc/src/setjmp/arm/setjmp.cpp:
```diff
@@ -0,0 +1,23 @@
+[[gnu::naked]]
+LLVM_LIBC_FUNCTION(int, setjmp, (__jmp_buf * buf)) {
+  asm(R"(
+      mov r12, sp
```
ditto

nickdesaulniers replied:
This code is saving r4-r11, sp, and lr registers to the memory referred to by `buf`, then returning zero.

nickdesaulniers replied:
added comments PTAL

nickdesaulniers commented:
@petrhosek mentioned to me offline the need to support `--target=armv6m-unknown-eabi` which the current implementation does not.  Let me update my implementation to add support for ARM or THUMB(1).

nickdesaulniers committed:
fix up setjmp for thumb1

nickdesaulniers committed:
save lr, add more comments

nickdesaulniers committed:
implement longjmp for thumb1

nickdesaulniers committed:
add thumb target fn attr to longjmp to match setjmp

SchrodingerZhu commented on libc/src/setjmp/arm/setjmp.cpp:
```diff
@@ -0,0 +1,59 @@
+
+#if defined(__thumb__) && __ARM_ARCH_ISA_THUMB == 1
+
+[[gnu::naked, gnu::target("thumb")]]
```
just curious, does `arm/aarch64` also support multi-versioned symbols?

smithp35 replied:
If multi-versioned symbols is referring to function multi-versioning then there is a beta version defined in https://arm-software.github.io/acle/main/acle.html#function-multi-versioning AFAIK it has been concentrating on AArch64 though.

Unlikely to be used in bare-metal cases like cortex-m0 though.

nickdesaulniers replied:
Right, depends on loader support, which we currently lack.  Also, not sure yet we _want_ to support a libc with mixed arm/thumb mode support.  Frankly, I'd suspect that folks that _care_ about libc support on 32b ARM only really care about Thumb mode (unless they're running on some variant with only ARM mode support).

smithp35 commented:
Apologies for the delay in responding, yes Thumb 1 only is important. There is also armv8-m.base (cortex-m23) which is the Arm v8 version of Thumb 1 only v6-m.

I've asked our C-library expert if he can take a look at the Arm specific details.

The main problem LLVM-libc will have v6-m is that it doesn't have any load/store exclusive instructions so it is unsuitable for multithreaded implementations. Strictly speaking interrupts can be disabled, but this is a privileged instruction.

statham-arm commented on libc/src/setjmp/arm/longjmp.cpp:
```diff
@@ -0,0 +1,68 @@
+      mov lr, r3
+
+      # return val ?: 1;
+      movs r0, r1
```
This use of r1 ought to be accessing the value passed in to `longjmp` as a parameter. But that's not what's in r1 right now. The previous two `ldmia` both clobbered it. At the point of this instruction, the value in r1 is whatever value the jmp_buf had saved for r11.

nickdesaulniers replied:
good catch! 1e28ecb25cde

smithp35 commented:
Not sure you're aware of it so I'll mention it anyway. There is some ABI documentation about requirements for Arm/Thumb setjmp/longjmp in the clib32 https://github.com/ARM-software/abi-aa/blob/main/clibabi32/clibabi32.rst#setjmph

smithp35 commented:
I know this is titled basic implementation, so these are likely some TODOs for a full implementation:
* Floating point/MVE registers. In all known implementations these are d0-d16.
* Accounting for security extensions like v8.1-m PACBTI, where the return address may be signed/authenticated, and BTI landing pads may be needed at certain points.

nickdesaulniers committed:
restore callee-saved r4, r5, r6 for thumb1, use r1-r6 as scratch rather than r2-r7

Link: https://github.com/ARM-software/abi-aa/blob/main/aapcs32/aapcs32.rst#core-registers

nickdesaulniers committed:
longjmp avoid clobbering r1, tail merge

nickdesaulniers commented on libc/src/setjmp/arm/setjmp.cpp:
```diff
@@ -0,0 +1,59 @@
+      # together with the rest.
+      mov r2, r8
+      mov r3, r9
+      mov r4, r10
```
good catch! 349e87430cfd

nickdesaulniers committed:
add some links to todos

nickdesaulniers commented:
> I know this is titled basic implementation, so these are likely some TODOs for a full implementation:
> 
> * Floating point/MVE registers. In all known implementations these are d0-d16.
> * Accounting for security extensions like v8.1-m PACBTI, where the return address may be signed/authenticated, and BTI landing pads may be needed at certain points.

Thanks for the tips! 6bed9ef91184 added some TODOs that I've filed.  As @frobtech notes, there's even more that we can be doing here, too.

Looking at Bionic's implementation, it quotes the AAPC32 to claim that only d8-d15 need be preserved?

https://github.com/ARM-software/abi-aa/blob/main/aapcs32/aapcs32.rst#vfp-register-usage-conventions

says d8-d15, q4-q7

nickdesaulniers commented:
right on, thanks for the review @statham-arm !!  Would you have cycles to help review other 32b ARM related patches I may have in the near future? https://github.com/llvm/llvm-project/issues/93738 comes to mind.  Also, I'll be on paternity leave for the next two weeks (on again, off again).

nickdesaulniers commented:
> Not sure you're aware of it so I'll mention it anyway. There is some ABI documentation about requirements for Arm/Thumb setjmp/longjmp in the clib32 https://github.com/ARM-software/abi-aa/blob/main/clibabi32/clibabi32.rst#setjmph

I was not! Thanks for the reference.  Here's some more TODO's I filed after reading that section:
- https://github.com/llvm/llvm-project/issues/94072
- https://github.com/llvm/llvm-project/issues/94071
- https://github.com/llvm/llvm-project/issues/94069

nickdesaulniers committed:
formatting

statham-arm commented on libc/src/setjmp/arm/setjmp.cpp:
```diff
@@ -0,0 +1,71 @@
+      # r4, r5, and r6. r7 and r8 were not clobbered. These register are 4B, so
+      # 10 registers times 4B gives us 40B to rewind buf by.
+      subs r0, r0, #40
+      ldmia r0!, {r4-r6}
```
I'm confident this is right, but I'm curious why you didn't do it the same way as in `longjmp`, by doing the original store in smaller blocks so that you don't clobber those registers in the first place. Set r1,r2,r3 to r8,r9,r10 and do a three-reg `stmia`; then set them to r11,sp,lr and do another one.

One fewer instruction and less memory access. It's _more_ of a win here to do it like this than it was in `longjmp`, because you have one more low register free, so you can split the store into two 3-reg chunks instead of three 2-reg ones.

nickdesaulniers replied:
as in:
```diff
diff --git a/libc/src/setjmp/arm/setjmp.cpp b/libc/src/setjmp/arm/setjmp.cpp
index aef1bed5e8e4..287e09c4a08b 100644
--- a/libc/src/setjmp/arm/setjmp.cpp
+++ b/libc/src/setjmp/arm/setjmp.cpp
@@ -26,19 +26,12 @@ LLVM_LIBC_FUNCTION(int, setjmp, (__jmp_buf * buf)) {
       mov r1, r8
       mov r2, r9
       mov r3, r10
-      mov r4, r11
-      mov r5, sp
-      mov r6, lr
-      stmia r0!, {r1-r6}
+      stmia r0!, {r1-r3}
 
-      # AAPCS32 states
-      # A subroutine must preserve the contents of the registers r4-r8 ...
-      # so rewind the buf pointer by the number of registers saved (i.e. 10
-      # registers: r4, r5, r6, r7, r8, r9, r10, r11, sp, lr), then restore
-      # r4, r5, and r6. r7 and r8 were not clobbered. These register are 4B, so
-      # 10 registers times 4B gives us 40B to rewind buf by.
-      subs r0, r0, #40
-      ldmia r0!, {r4-r6}
+      mov r1, r11
+      mov r2, sp
+      mov r3, lr
+      stmia r0!, {r1-r3}
 
       # Return 0.
       movs r0, #0
```
?

nickdesaulniers replied:
5dac5006c692

statham-arm replied:
Yes, that's exactly what I meant. Much nicer!

statham-arm reviewed (approved):
LGTM. I have a minor performance suggestion about the Thumb-1 `setjmp`, but I won't insist on it.

nickdesaulniers committed:
more consistent use of - in {} reglists

nickdesaulniers committed:
avoid r4-r6 reload

nickdesaulniers commented:
Thanks again for the reviews @statham-arm and @frobtech !
